projects/pgai/pgai/vectorizer/worker_tracking/worker_tracking.py

```diff
@@ -0,0 +1,140 @@
+import asyncio
+import datetime
+from uuid import UUID
+
+import psycopg
+import structlog
+
+from ..features import Features
+
+log = structlog.get_logger()
+
+
+class WorkerTracking:
+    def __init__(
+        self, db_url: str, poll_interval: int, features: Features, version: str
+    ):
+        self.db_url = db_url
+        self.poll_interval = poll_interval
+        self.worker_id: UUID | None = None
+        self.num_errors_since_last_heartbeat = 0
+        self.error_message = None
+        self.enabled = features.worker_tracking
+        self.version = version
+        self.num_successes_since_last_heartbeat = 0
+
+    async def start(self) -> None:
+        if not self.enabled:
+            return
+
+        async with (
+            await psycopg.AsyncConnection.connect(self.db_url, autocommit=True) as conn,
+            conn.cursor() as cur,
+            conn.transaction(),
+        ):
+            poll_interval_td = datetime.timedelta(seconds=self.poll_interval)
+            await cur.execute(
+                "select ai._worker_start(%s::text, %s::interval)",
+                (self.version, poll_interval_td),
+            )
+            res = await cur.fetchone()
+            if res is None:
+                raise Exception("Failed to start worker tracking")
+            self.worker_id = res[0]
+
+        asyncio.create_task(self.heartbeat())
+
+    def get_short_worker_id(self) -> str:
+        # Return first segment of UUID (before first hyphen)
+        return str(self.worker_id).split("-")[0] if self.worker_id else ""
+
+    async def _report_error(self, error_message: str) -> None:
+        self.num_errors_since_last_heartbeat += 1
+        self.error_message = error_message
+
+    async def force_heartbeat(self) -> None:
+        if not self.enabled:
+            return
+
+        async with await psycopg.AsyncConnection.connect(
+            self.db_url,
+            autocommit=True,
+            application_name=f"pgai-worker-heartbeat-force: {self.get_short_worker_id()}",  # noqa: E501
+        ) as conn:
+            await self._heartbeat(conn)
+
+    async def _heartbeat(self, conn: psycopg.AsyncConnection) -> None:
+        async with conn.cursor() as cur, conn.transaction():
+            num_errors = self.num_errors_since_last_heartbeat
+            self.num_errors_since_last_heartbeat = 0
+            error_message = self.error_message
+            self.error_message = None
+            num_successes = self.num_successes_since_last_heartbeat
+            self.num_successes_since_last_heartbeat = 0
+            await cur.execute(
+                "select ai._worker_heartbeat(%s, %s, %s, %s)",
+                (self.worker_id, num_successes, num_errors, error_message),
+            )
+
+    async def heartbeat(self) -> None:
+        if not self.enabled:
+            return
+
+        failures = 0
+        while failures < 3:
```

Comment:
A message informing we met this condition might be useful, we can rely on `heartbeat failed` message count because it gets restarted.

Reply:
done